src/components/form/adapters/base-adapter.ts

```diff
@@ -0,0 +1,211 @@
+        // The events to bind the the web component
+        events: {},
+        // The props to pass to the web component
+        elementProps: {},
+    };
+
+    constructor(
+        public props: {
+            name: string;
+            events?: any;
+            elementProps?: any;
+        }
+    ) {
+        super(props);
+
+        this.state = {
+            prevEvents: {},
+        };
+
+        this.component = React.createRef();
+
+        this.setComponentProperty = this.setComponentProperty.bind(this);
+    }
+
+    // Sets a property on the web component
+    setComponentProperty([key, value]: [string, any]) {
+        const element = this.component.current;
+        element[key] = value;
+    }
+
+    /**
+     * Get the handler for an event.
+     *
+     * The handler stop propagation otherwise. Should be used instead of just binding
+     * the handler passed in the `events` prop due to this reason
+     *
+     * @param {string} eventName the name of the event to get the handler for
+     *
+     * @returns {any} the handler to bind to the component for the event name
+     */
+    getHandler(eventName: string) {
+        const { events } = this.props;
+        return (event) => {
+            event.stopPropagation();
+            events[eventName](event);
+        };
+    }
+
+    /**
+     * Set events on the web component
+     *
+     * @param {object} events Event names to handlers
+     *
+     * @returns {void}
+     */
+    setEvents(events: object) {
+        const element = this.component.current;
+        for (const event of Object.keys(events)) {
+            element.addEventListener(event, this.getHandler(event));
+        }
+    }
+
+    /**
+     * Remove events on the web component
+     *
+     * @param {object} events Event names to handlers
+     *
+     * @returns {void}
+     */
+    removeEvents(events: object) {
+        const element = this.component.current;
+        for (const event of Object.keys(events)) {
+            element.removeEventListener(event, events[event]);
+        }
+    }
+
+    /**
+     * Detect if event handlers have changed. If they
+     * have changed, store them in the state, and remove and
+     * re-add the changed event handlers
+     *
+     * @returns {void}
+     */
+    updateEvents() {
+        const { prevEvents } = this.state;
+        const { events } = this.props;
+
+        const eventsToAdd = {};
+        const eventsToRemove = {};
+        let eventsChanged = false;
+        for (const eventName of Object.keys(events)) {
+            if (prevEvents[eventName] !== events[eventName]) {
+                eventsToAdd[eventName] = events[eventName];
+                eventsToRemove[eventName] = prevEvents[eventName];
+                eventsChanged = true;
+            }
+        }
+
+        if (eventsChanged) {
+            this.setState({ prevEvents: events });
+        }
+
+        this.removeEvents(eventsToRemove);
+        this.setEvents(eventsToAdd);
```

Comment:
Do we still need to do all this even if nothing has changed?

Reply:
we do not